CRM/Activity/Import/Parser/Activity.php

```diff
@@ -182,61 +106,63 @@ protected function getRequiredFields(): array {
+        + (array) Activity::getFields()
+          ->addWhere('readonly', '=', FALSE)
+          ->addWhere('usage', 'CONTAINS', 'import')
+          ->setAction('save')
+          ->addOrderBy('title')
+          ->execute()->indexBy('name');
+      $contactFields = $this->getContactFields('Individual');
+      foreach ($contactFields as &$field) {
+        // At this stage all fields except those specifically marked are the target_contact.
+        $field['entity_instance'] = 'TargetContact';
       }
-      $tmpContactField['external_identifier'] = $contactFields['external_identifier'];
-      $tmpContactField['external_identifier']['title'] = $contactFields['external_identifier']['title'] . ' (target contact)' . ' (match to contact)';
-      $tmpContactField['source_contact_external_identifier'] = $contactFields['external_identifier'];
-      $tmpContactField['source_contact_external_identifier']['title'] = $contactFields['external_identifier']['title'] . ' (source contact)' . ' (match to contact)';
-
-      $fields = array_merge($fields, $tmpContactField);
-      $fields = array_merge($fields, $tmpFields);
-      $fields = array_merge($fields, CRM_Core_BAO_CustomField::getFieldsForImport('Activity'));
-
-      $fields = array_merge($fields, [
-        'source_contact_id' => [
-          'title' => ts('Source Contact'),
-          'headerPattern' => '/Source.Contact?/i',
-          'name' => 'source_type_id',
-          'options' => FALSE,
-          'type' => CRM_Utils_Type::T_INT,
-        ],
-        'target_contact_id' => [
-          'title' => ts('Target Contact'),
-          'headerPattern' => '/Target.Contact?/i',
-          'name' => 'target_type_id',
-          'options' => FALSE,
-          'type' => CRM_Utils_Type::T_INT,
-        ],
-      ]);
+      $fields['target_contact.id'] = $contactFields['id'];
+      $fields['target_contact.id']['entity'] = 'Contact';
+      $fields['target_contact.id']['match_rule'] = '*';
+      $fields['source_contact.id'] = $fields['target_contact.id'];
+      $fields['source_contact.id']['entity_instance'] = 'SourceContact';
+      $fields['source_contact.id']['title'] .= ' ' . ts('(match to source contact)');
+      $fields['target_contact.id']['title'] .= ' ' . ts('(match to target contact)');
+      unset($contactFields['id']);
+
+      $fields += $contactFields;
+      $fields['target_contact.external_identifier'] = $contactFields['external_identifier'];
+      $fields['target_contact.external_identifier']['title'] = $contactFields['external_identifier']['title'] . ' (target contact)';
+      $fields['source_contact.external_identifier'] = $contactFields['external_identifier'];
+      $fields['source_contact.external_identifier']['entity_instance'] = 'SourceContact';
+      $fields['source_contact.external_identifier']['title'] = $contactFields['external_identifier']['title'] . ' (source contact)';
+      unset($fields['external_identifier']);
```

Comment:
In Api4 generally the these pseudo-fields end in `_id`. Not sure if important to keep that consistent here.

https://github.com/civicrm/civicrm-core/blob/ddcc28ae4c8829f62b8c1e5d49478d7eb6b2a6d5/Civi/Api4/Service/Spec/Provider/ActivitySpecProvider.php#L52-L85

Reply:
@colemanw no - a feature of the importer is that it handles a mix of things on those fields - ids, names, labels, abbreviations with flexibility around case sensitivity. I know you won't like that :-) but it is not only because of legacy code - it really is the case that users are pretty sloppy with their thinking about the differences between those things